File: cargo/crates/numbers-utils/tests/operations/fibonacci_utils_tests.rs
use chrono::Local;
use colored::Colorize;

use num_bigint::BigUint;
use numbers_utils::operations::fibonacci_utils::{fibonacci, fibonacci_random};

fn format_error_message(expected: &BigUint, actual: &BigUint) -> colored::ColoredString {
    format!(
        "\n\nTest run at: {}\nExpected : {}\nActual   : {}\n",
        Local::now().to_rfc3339(),
        expected,
        actual
    )
    .red()
}

#[test]
fn fibonacci_test() {
    let n = 100;
    let expected_output = BigUint::parse_bytes(b"354224848179261915075", 10).unwrap();
    let actual_output = fibonacci(n);
    let error_message = format_error_message(&expected_output, &actual_output);
    assert_eq!(expected_output, actual_output, "{}", error_message);
}

#[test]
fn fibonacci_random_test() {
    let result = fibonacci_random();
    let start_string = "Fib[";
    let error_message = format!("\nExpected: {}\n     Got: {}\n", start_string, result).red();
    assert!(result.starts_with(start_string), "{}", error_message,);
    assert!(result.contains(" = "));
}

File: gradle/subprojects/numbers-utils/src/test/kotlin/com/app/numbers/utils/operations/FibonacciUtilsTests.kt
package com.app.numbers.utils.operations

import org.apache.commons.lang3.time.DateFormatUtils
import java.math.BigInteger
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertTrue

class FibonacciUtilsTests {
    private fun formatErrorMessage(
        expected: BigInteger,
        actual: BigInteger,
    ): String {
        val now = System.currentTimeMillis()
        val formattedDate = DateFormatUtils.format(now, "yyyy-MM-dd'T'HH:mm:ss")
        return """
            |${"\u001B[31m"}
            |Test run at: $formattedDate
            |Expected : $expected
            |  Actual : $actual
            |
            |${"\u001B[00m"}
            """.trimMargin()
    }

    @Test
    fun `fibonacci test`() {
        val n = 100
        val expectedOutput = BigInteger("354224848179261915075")
        val actualOutput = FibonacciUtils.fibonacci(n)
        val errorMessage = formatErrorMessage(expectedOutput, actualOutput)
        assertEquals(expectedOutput, actualOutput, errorMessage)
    }

    @Test
    fun `fibonacci random test`() {
        val result = FibonacciUtils.fibonacciRandom()
        val startString = "Fib["
        val errorMessage =
            """
            |${"\u001B[31m"}
            |Expected: $startString
            |     Got: $result${"\u001B[00m"}
            """.trimMargin()
        assertTrue(result.startsWith(startString), errorMessage)
        assertTrue(result.contains(" = "))
    }
}
